src/mpid/ch4/netmod/ofi/ofi_events.h

```diff
@@ -138,8 +150,25 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_recv_complete(MPIR_Request * rreq, int ev
         MPI_Status *status = &rreq->status;
         MPIR_Request *am_req = MPIDI_OFI_REQUEST(rreq, am_req);
         int am_recv_id = MPIDI_OFI_REQUEST(rreq, am_handler_id);
-        mpi_errno = MPIDIG_global.tag_recv_cbs[am_recv_id] (am_req, status);
-        MPIR_ERR_CHECK(mpi_errno);
+        if (am_recv_id == -1) {
+            /* native rndv direct */
+
+            /* copy COUNT and MPI_ERROR, but skip MPI_SOURCE and MPI_RANK */
+            am_req->status.count_lo = rreq->status.count_lo;
+            am_req->status.count_hi_and_cancelled = rreq->status.count_hi_and_cancelled;
+            am_req->status.MPI_ERROR = rreq->status.MPI_ERROR;
```

Comment:
who owns `am_req` in this case? I'm confused why the "direct" method involves a separate am request.

Reply:
All internal messages including pipeline chunks and control messages use a separately allocated "ofi" request rather than reuse the original mpi request. This "am_req" is the original mpi request.

It is named "am_req" because we are reusing `MPIDI_NM_am_tag_send/recv` for direct send and `MPIDI_NM_am_tag_send/recv` was originally designed for MPIDIG layer to send data natively. It is a confusing name now that we are double-using `MPIDI_NM_am_tag_send/recv`. We should clear it up in a future PR.

Reply:
OK, I think I better understand what "direct" is now. Agreed we can work on naming issues later on.